runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/net/IpAddr.kt

```diff
@@ -0,0 +1,416 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package aws.smithy.kotlin.runtime.net
+
+import aws.smithy.kotlin.runtime.util.InternalApi
+
+/**
+ * An IP Address (either IPv4 or IPv6)
+ */
+@InternalApi
+public sealed class IpAddr {
+    public companion object {
+        /**
+         * Parse a string into an [IpAddr]. Fails if [s] is not a valid IP address
+         */
+        public fun parse(s: String): IpAddr = when {
+            s.isIpv4() -> Ipv4.parse(s)
+            else -> Ipv6.parse(s)
+        }
+    }
+
+    /**
+     * The raw numerical address
+     */
+    public abstract val octets: ByteArray
+
+    /**
+     * The resolved numerical address represented as a string
+     */
+    public abstract val address: String
+
+    /**
+     * Returns true if this is the loopback address
+     */
+    public abstract val isLoopBack: Boolean
+
+    /**
+     * True if this is an [Ipv4] instance
+     */
+    public val isIpv4: Boolean
+        get() = this is Ipv4
+
+    /**
+     * True if this is an [Ipv6] instance
+     */
+    public val isIpv6: Boolean
+        get() = this is Ipv6
+
+    /**
+     * An IPv4 address as defined by [RFC 791](https://www.rfc-editor.org/rfc/rfc791)
+     */
+    public data class Ipv4(
```

Comment:
**bikeshed**

`IpAddr.Ipv4` or `IpAddr.V4`?

Reply:
> I think I'm leaving towards changing this to  IpAddr.V4...

Reply:
Oh good idea. 👍

Reply:
Well if we break out the variants into separate files then I'm less inclined to name it just `V4` or `V6` because it won't be namespaced by `IpAddr` anymore. 

Let's simplify with some real alternatives to consider:

ALT1:
```kotlin
sealed class IpAddr {
    class V4(...): IpAddr()

    class V6(...): IpAddr()
}
```

usage
```kotlin
val v4Addr = IpAddr.V4(...)
```
-----

ALT2
```kotlin
sealed class IpAddr {
     // common stuff
}

// file: IpV4Addr.kt
class IpV4Addr(...): IpAddr()


// file: IpV6Addr.kt
class IpV6Addr(...): IpAddr()

```

usage
```kotlin
val v4Addr = IpV4Addr(...)
```